docbrown/core/src/tgraph_shard.rs

```diff
@@ -85,23 +116,59 @@ impl TGraphShard {
     }
 
     #[inline(always)]
-    pub fn write_shard<A, F>(&self, f: F) -> A
+    fn write_shard<A, F>(&self, f: F) -> A
     where
         F: FnOnce(&mut TemporalGraph) -> A,
     {
-        let mut shard = self.rc.write();
+        let mut binding = self.rc.write();
+        let mut shard = binding.as_mut().unwrap();
```

Comment:
I'd like this to change to `Result<A, GraphError>`
but I'd like you do make some performance tests before and see what impact does it have 
to support the error `Attempting to write to frozen graph`

Use this crate 
https://docs.rs/thiserror/latest/thiserror/